kolena/workflow/metrics/_plots.py

```diff
@@ -0,0 +1,119 @@
+# Copyright 2021-2023 Kolena Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from collections import defaultdict
+from typing import Dict
+from typing import List
+from typing import Tuple
+from typing import Union
+
+import numpy as np
+from sklearn.metrics import precision_recall_fscore_support
+
+from kolena.workflow.evaluator import Curve
+from kolena.workflow.evaluator import CurvePlot
+from kolena.workflow.metrics._geometry import InferenceMatches
+from kolena.workflow.metrics._geometry import MulticlassInferenceMatches
+
+
+def _compute_sklearn_arrays(
+    all_matches: List[Union[MulticlassInferenceMatches, InferenceMatches]],
+) -> Tuple[List[int], List[int], Dict[str, Tuple[List[int], List[int]]]]:
+    arrays_by_label: defaultdict[str, Tuple[List[int], List[int]]] = defaultdict(lambda: ([], []))
+    y_true: List[int] = []
+    y_score: List[int] = []
+    for image_bbox_matches in all_matches:
+        for _, bbox_inf in image_bbox_matches.matched:  # TN (if above threshold)
+            arrays_by_label[bbox_inf.label][0].append(1)
+            arrays_by_label[bbox_inf.label][1].append(bbox_inf.score)
+            y_true.append(1)
+            y_score.append(bbox_inf.score)
+        for bbox_gt, _ in image_bbox_matches.unmatched_gt:  # FN
+            y_true.append(1)
+            y_score.append(-1)
+            arrays_by_label[bbox_gt.label][0].append(1)
+            arrays_by_label[bbox_gt.label][1].append(-1)
+        for bbox_inf in image_bbox_matches.unmatched_inf:  # FP (if above threshold)
+            arrays_by_label[bbox_inf.label][0].append(0)
+            arrays_by_label[bbox_inf.label][1].append(bbox_inf.score)
+            y_true.append(0)
+            y_score.append(bbox_inf.score)
+    return y_true, y_score, dict(arrays_by_label)
+
+
+def _compute_threshold_curves(
+    y_true: np.ndarray,
+    y_score: np.ndarray,
+    label: str,
+) -> Tuple[Curve, Curve]:
+    thresholds = list(np.linspace(min(abs(y_score)), max(y_score), min(401, len(y_score))))[:-1]
```

Comment:
Curious about this piece:
```
min(401, len(y_score))
```
`len(y_score)` is strange to linspace with -- if you're going to have a point for every y_score, why not just use the threshold values (instead of linspacing)?

On the flip side, one consideration that we had for builtin OD was that linspacing gave us more points when values were sparse, but less detail than we would have liked when values were dense. Not a blocker, but worth pointing out that you could have 100 points between (0.9525, 0.9550) and only 10 points between (0, 0.5) -- linspacing smudges that out.

Last thing to point out -- there's a chance that you're missing the f1 optimal point if you linspace. Do we want to ensure that it gets surfaced?

Reply:
I made more progress on the entire workflow for a better understanding of the [full picture](https://github.com/kolenaIO/kolena/pull/56#discussion_r1214373355). Here, plot utils do not have to be at the test case level, but for the OD workflow, they will be per test case. I think a good solution to address missing the optimal threshold (for sparse cases) is doubling the size of the linspace.

Reply:
@y27choi is doubling the size of the linspace from `min(401, len(y_score))` to `min(401, 2 * len(y_score))` a good idea?